### PR TITLE
[SPARK-37294][SQL][TESTS] Check inserting of ANSI intervals into a table partitioned by the interval columns

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Add new tests to check that ANSI intervals are supported by `INSERT INTO` for tables partitioned by the interval columns. New tests check v1/v2 In-Memory catalog and static/dynamic overwrite.

### Why are the changes needed?
To improve test coverage.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *DataSourceV2SQLSuite"
$ build/sbt "test:testOnly *InsertSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @sarutak @AngersZhuuuu @Peng-Lei Could you review this PR, please.

**Review by HyukjinKwon — APPROVED** [review]

**Review by AngersZhuuuu — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 1]
```diff
@@ -2915,6 +2915,40 @@ class DataSourceV2SQLSuite
...
 
+  test("SPARK-37294: insert ANSI intervals into a table partitioned by the interval columns") {
+    val tbl = "testpart.interval_table"
+    Seq(PartitionOverwriteMode.DYNAMIC.toString,
+      PartitionOverwriteMode.STATIC.toString).foreach { mode =>
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> mode) {
+        withTable(tbl) {
+          sql(
+            s"""
+               |CREATE TABLE $tbl (i int, part1 INTERVAL YEAR, part2 INTERVAL DAY) USING $v2Format
```
Comment (AngersZhuuuu) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2926]: i Int

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 2]
```diff
@@ -2915,6 +2915,40 @@ class DataSourceV2SQLSuite
...
+            s"""
+               |INSERT INTO $tbl
+               | PARTITION (part1 = INTERVAL '8' YEAR, part2 = INTERVAL '9' DAY)
+               |SELECT 7""".stripMargin)
+
+          checkAnswer(
+            spark.table(tbl),
+            Seq(Row(1, java.time.Period.ofYears(2), java.time.Duration.ofDays(3)),
+              Row(4, java.time.Period.ofYears(5), java.time.Duration.ofDays(6)),
+              Row(7, java.time.Period.ofYears(8), java.time.Duration.ofDays(9))))
```
Comment (AngersZhuuuu) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2946]: import jave.time.Period & java.time.Duration?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 3]
```diff
@@ -1063,6 +1063,40 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+            s"""
+               |INSERT INTO $tbl
+               | PARTITION (part1 = INTERVAL '8' YEAR, part2 = INTERVAL '9' DAY)
+               |SELECT 7""".stripMargin)
+
+          checkAnswer(
+            spark.table(tbl),
+            Seq(Row(1, java.time.Period.ofYears(2), java.time.Duration.ofDays(3)),
+              Row(4, java.time.Period.ofYears(5), java.time.Duration.ofDays(6)),
+              Row(7, java.time.Period.ofYears(8), java.time.Duration.ofDays(9))))
```
Comment (AngersZhuuuu) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1095]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 1 above)
Comment (sarutak) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2926]: `i INT`?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 4]
```diff
@@ -1063,6 +1063,40 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+
+  test("SPARK-37294: insert ANSI intervals into a table partitioned by the interval columns") {
+    val tbl = "interval_table"
+    Seq(PartitionOverwriteMode.DYNAMIC.toString,
+      PartitionOverwriteMode.STATIC.toString).foreach { mode =>
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> mode) {
+        withTable(tbl) {
+          sql(
+            s"""
+              |CREATE TABLE $tbl (i int, part1 INTERVAL YEAR, part2 INTERVAL DAY) USING PARQUET
```
Comment (sarutak) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1075]: ditto.

**Review by sarutak — APPROVED** [review]
Left minor comments. Otherwise LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2926]: > `i INT`?

Ah. yea

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 5]
```diff
@@ -2915,6 +2915,40 @@ class DataSourceV2SQLSuite
     }
   }
 
+  test("SPARK-37294: insert ANSI intervals into a table partitioned by the interval columns") {
+    val tbl = "testpart.interval_table"
+    Seq(PartitionOverwriteMode.DYNAMIC.toString,
+      PartitionOverwriteMode.STATIC.toString).foreach { mode =>
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2921]: nit. indentation.

**Comment by Peng-Lei** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 6]
```diff
@@ -2915,6 +2915,39 @@ class DataSourceV2SQLSuite
...
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> mode.toString) {
+        withTable(tbl) {
+          sql(
+            s"""
+               |CREATE TABLE $tbl (i INT, part1 INTERVAL YEAR, part2 INTERVAL DAY) USING $v2Format
+               |PARTITIONED BY (part1, part2)
+              """.stripMargin)
+
+          sql(
+            s"""ALTER TABLE $tbl ADD PARTITION (
```
Comment (Peng-Lei) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2930]: nit:
s"""
     |ALTER TABLE $tbl ADD PARTITION (

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @Peng-Lei @sarutak @HyukjinKwon @dongjoon-hyun @cloud-fan @AngersZhuuuu for review.

**Pull request closed without merging**